test-data/unit/check-functions.test

```diff
@@ -1475,6 +1475,17 @@ else:
     @dec
     def f(): pass
 
+[case testConditionalFunctionDefinitionWithTuple]
+def bar() -> None:
+    # weirdly, needs to be an empty tuple. empty lists and various other special assignments
+    # don't trigger the bug
+    cond = ()
+    if cond:
```

Comment:
A simple test would be just `if False:` The problem is that mypy considers if branch unreachable, and therefore doesn't type check it, and therefore the original definition doesn't get its type inferred (and stays `None`).

I think a better solution is to simply not show any error in this case:
* It will be more consistent with not showing error in unreachable code
* Technically, there is no _redefinition_, only one definition is ever executed.

Reply:
@ilevkivskyi for what it's worth, the crash does not repro with `if False` and several other variations I tried